Fixed the `Bosh::AzureCloud::Cloud.initialize` method's signature so …

Justin-W opened the pull request:
…that the 2nd argument is optional.

Note: Based on the changes made in commit 6c6ec5909b3635d5ea14a5f4dfa2263f2edf1468 (which added the 2nd argument), it seems that the argument was intended to be optional from the start. However, according to the method's signature, it was technically a required argument (despite that the method handles nil/missing values internally on line 29). Also, the following code already doesn't pass the 2nd argument:
- https://github.com/cloudfoundry/bosh-azure-cpi-release/blob/9754cf84f18be7ce0250518bd5a660a459470b56/src/bosh_azure_cpi/bin/bosh_azure_console#L29

Note: The separate handling of the argument's default value on lines 26 and 29 are not redundant. The default value in the signature (line 26) only applies when the optional arg is omitted by the caller. If the caller explicitly specifies the argument with a value of `nil`, the arg default doesn't override the nil, but the `|| 1` on line 29 will. And without the argument default on line 26, the argument technically isn't optional, it is required. Meaning that it is technically an error for any caller to omit that argument, and that line 29 would therefore be covering up the caller's misuse of the method. So both "defaults" are needed in order to correctly handle all of the possible method invocation options for that argument.

Here is the error message which was previously being detected (prior to this change):
> Missing argument 'api_version'. Required 2, 'options, api_version'

# Checklist:

Please check each of the boxes below for which you have completed the corresponding task:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ N/A ] I have added tests that prove my fix is effective or that my feature works
- [x] All unit tests pass locally (after my changes)
- [x] Rubocop reports zero offenses (after my changes)

Please include below the summary portions of the output from the following 2 scripts:

  ```
  pushd src/bosh_azure_cpi
    ./bin/test-unit
    ./bin/rubocop_check
  popd
  ```

  _NOTE:_ Please see how to setup dev environment and run unit tests in docs/development.md.

### Unit Test output:

> 972 examples, 0 failures
> Coverage report generated for RSpec to /Users/justinw/go/src/github.com/cloudfoundry/bosh-azure-cpi-release/src/bosh_azure_cpi/coverage. 4182 / 4203 LOC (99.5%) covered.

### Rubocop output:

> 184 files inspected, no offenses detected

# Changelog

* Changed the `Bosh::AzureCloud::Cloud.initialize` method's signature to make the 2nd positional argument explicitly optional

bgandon commented:
I think we should not turn the second parameter as having any default value if it never had, and without any requirement for it.

We should also not use any silent defaults that fix mis-use of the code, and also not implement unnecessary defensive measures. So, just conform to the _let-it-crash_ principle.

Let's keep it simple, please only remove the `|| 1` [in `@api_version = api_version || 1`](https://github.com/cloudfoundry/bosh-azure-cpi-release/blob/a283bed/src/bosh_azure_cpi/lib/cloud/azure/cloud.rb#L28), update [the incorrect calling from `bin/bosh_azure_console` script](https://github.com/cloudfoundry/bosh-azure-cpi-release/blob/a283bed50d/src/bosh_azure_cpi/bin/bosh_azure_console#L29), using [the current API version](https://github.com/cloudfoundry/bosh-azure-cpi-release/blob/a283bed/src/bosh_azure_cpi/lib/cloud/azure/cloud.rb#L15), then add some test for better code coverage. So that in the future, some test can detect the issue you've pointed here. Thanks.

Justin-W commented:
@bgandon I've revised the PR as you requested:
1. the method now conforms to the _let-it-crash_ principle, raises an error if any invalid value is passed for the argument
2. updated all invocations which were lacking an explicit value for that arg
3. Updated the method's unit specs to test all of the valid invocation options and also the most likely invalid ones

bgandon reviewed (approved):
Okay 😅, I think that's a bit overkill amount of code for the severity of the initial concern to be honest, but I appreciate the tenacity of your commitment in making it right 👍
LGTM.

rkoster reviewed (approved):
LGTM

rkoster commented:
Thanks! @Justin-W

rkoster commented:
Just for context, this PR has broken some behavior. See https://github.com/cloudfoundry/bosh-azure-cpi-release/commit/f1f495c25dba8376c61a514604bc67c816fb7c8a for more details.
